core/helpers.py

```diff
@@ -58,3 +62,56 @@ def handle_cms_response(response):
         raise Http404()
     response.raise_for_status()
     return response.json()
+
+
+class GeoLocationRedirector:
+    DOMESTIC_COUNTRY_CODES = ['GB', 'IE']
+    COUNTRY_TO_LANGUAGE_MAP = {
+        'CN': 'zh-hans',
+        'DE': 'de',
+        'ES': 'es',
+        'JP': 'ja',
+    }
+    COOKIE_NAME = 'disable_geoloaction'
+    LANGUAGE_PARAM = 'lang'
+
+    def __init__(self, request):
+        self.request = request
+
+    @cached_property
+    def country_code(self):
+        client_ip, is_routable = get_client_ip(self.request)
+        if client_ip and is_routable:
+            response = GeoIP2().country(client_ip)
+            return response['country_code']
+
+    @property
+    def country_language(self):
+        return self.COUNTRY_TO_LANGUAGE_MAP.get(
+            self.country_code, settings.LANGUAGE_CODE
+        )
+
+    @property
+    def should_redirect(self):
+        return (
+            self.COOKIE_NAME not in self.request.COOKIES and
+            self.LANGUAGE_PARAM not in self.request.GET and
+            self.country_code is not None and
+            self.country_code not in self.DOMESTIC_COUNTRY_CODES
+        )
+
+    def get_response(self):
+        url = '{url}?{param}={lang}'.format(
```

Comment:
what if URL already has some query params set, e.g. UTM? This should check if query params are already present and append accordingly. We already do this somewhere with urlparse I believe, FAS or SSO.